hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/AdjustmentUtils.java

```diff
@@ -0,0 +1,232 @@
+     * @param feeCollector The fee collector
+     * @param allCollectorsAreExempt Whether all collectors are exempt
+     * @return The {@link FixedFee} representation of custom fee
+     */
+    public static CustomFee asFixedFee(
+            final long unitsToCollect,
+            final TokenID tokenDenomination,
+            final AccountID feeCollector,
+            final boolean allCollectorsAreExempt) {
+        Objects.requireNonNull(feeCollector);
+        final var spec = FixedFee.newBuilder()
+                .denominatingTokenId(tokenDenomination)
+                .amount(unitsToCollect)
+                .build();
+        return CustomFee.newBuilder()
+                .fixedFee(spec)
+                .feeCollectorAccountId(feeCollector)
+                .allCollectorsAreExempt(allCollectorsAreExempt)
+                .build();
+    }
+
+    /**
+     * Adjusts a HTS fee. If the fee is self-denominated, it should not trigger custom fees again
+     * So add the adjustment to previous level transaction. If the fee is not self-denominated,
+     * it should trigger custom fees again. So add the adjustment to next level transaction.
+     * @param result The {@link AssessmentResult} object
+     * @param sender The sender account
+     * @param collector The fee collector
+     * @param chargingTokenMeta The {@link CustomFeeMeta} object of token to be charged
+     * @param amount The amount to be charged
+     * @param denominatingToken The token denomination
+     */
+    public static void adjustHtsFees(
+            final AssessmentResult result,
+            final AccountID sender,
+            final AccountID collector,
+            final CustomFeeMeta chargingTokenMeta,
+            final long amount,
+            final TokenID denominatingToken) {
+        final var newHtsAdjustments = result.getHtsAdjustments();
+        final var inputHtsAdjustments = result.getInputTokenAdjustments();
+
+        // If the fee is self-denominated, we don't need tit to trigger next level custom fees
+        // So add assessments in given input transaction body.
+        if (chargingTokenMeta.tokenId().equals(denominatingToken)) {
+            // If the fee is self-denominated, it should not trigger custom fees again
+            // So add the adjustment to previous level transaction
+            result.addToExemptDebits(denominatingToken);
+            addHtsAdjustment(inputHtsAdjustments, sender, collector, amount, denominatingToken);
+        } else {
+            // Any change that might trigger next level custom fees should be added to next
+            // level transaction adjustments
+            addHtsAdjustment(newHtsAdjustments, sender, collector, amount, denominatingToken);
+        }
+    }
+
+    /**
+     * Adds HTS adjustment to given adjustments map. It makes 2 adjustments a debit
+     * for sender and credit to collector
+     * If there is already an entry merges the new change with it, otherwise creates a new entry.
```

Comment:
There _must_ already be an entry in the map, right? (Because self-denominated fees are by definition "re-using" the asset type being exchanged to charge fees.)

Reply:
This same method is used for adding new hts fees for non self denominated hts fixed fee 😃